storage/merge.go

```diff
@@ -458,47 +462,59 @@ func (c *chainSampleIterator) Seek(t int64) bool {
 			heap.Push(&c.h, iter)
 		}
 	}
-	return len(c.h) > 0
+	if len(c.h) > 0 {
+		c.curr = heap.Pop(&c.h).(chunkenc.Iterator)
+		return true
+	}
+	return false
 }
 
 func (c *chainSampleIterator) At() (t int64, v float64) {
-	if len(c.h) == 0 {
-		panic("chainSampleIterator.At() called after .Next() returned false.")
+	if c.curr == nil {
+		panic("chainSampleIterator.At() called before first .Next() or after .Next() returned false.")
 	}
-
-	return c.h[0].At()
+	return c.curr.At()
 }
 
 func (c *chainSampleIterator) Next() bool {
 	if c.h == nil {
-		for _, iter := range c.iterators {
+		c.h = samplesIteratorHeap{}
+		c.curr = c.iterators[0]
+		for _, iter := range c.iterators[1:] {
 			if iter.Next() {
 				heap.Push(&c.h, iter)
 			}
 		}
-
-		return len(c.h) > 0
 	}
 
-	if len(c.h) == 0 {
-		return false
-	}
-
-	currt, _ := c.At()
-	for len(c.h) > 0 {
-		nextt, _ := c.h[0].At()
-		// All but one of the overlapping samples will be dropped.
-		if nextt != currt {
-			break
+	for {
+		if c.curr != nil && c.curr.Next() {
+			currt, _ := c.curr.At()
+			if currt == c.lastt {
+				// Ignoring sample for the same timestamp.
+				continue
+			}
+			if len(c.h) == 0 {
+				return true
+			}
+			nextt, _ := c.h[0].At()
+			if currt < nextt {
+				c.lastt = currt
+				return true
+			}
+			heap.Push(&c.h, c.curr)
+		} else if len(c.h) == 0 {
+			c.curr = nil
+			return false
 		}
 
-		iter := heap.Pop(&c.h).(chunkenc.Iterator)
-		if iter.Next() {
-			heap.Push(&c.h, iter)
+		c.curr = heap.Pop(&c.h).(chunkenc.Iterator)
+		currt, _ := c.curr.At()
+		if currt != c.lastt {
```

Comment:
If we've gotten here, isn't this always going to be true?

Reply:
We come here when either curr is exhausted or it has crossed the top of the heap. It is possible that the top of the heap has the same timestamp as the lastt.